### ISSUE TITLE
[Bug] Cannot use string type literal value for jcr long type property

### ISSUE THREAD

**Issue by steffenbrem** [doctrine/phpcr-odm#688]
I have a document, where the `workspace` property is mapped as a `LONG`. When I execute the following query, it builds the wrong query, and returns zero rows (while I have 4 documents matching those criteria):
```php
$queryBuilder
    ->andWhere()
        ->eq()
            ->field('o.workspace')
            ->literal($value) // $value has a value of '1' (string)
;
```
This is the query statement it produces:
```
SELECT * FROM [nt:unstructured] AS o WHERE (o.[workspace] = '1' AND (o.[phpcr:class] =
'Mango\Component\Content\Model\Page' OR o.[phpcr:classparents] =
'Mango\Component\Content\Model\Page'))
```

Now, when I type cast the `$value` variable to be an int, it builds the proper query statement and casts `'1'` to `LONG`.

I think this is a bug, because you can pass in strings in the query builder for boolean type properties and those work fine.

I am using Jackrabbit 2 as my storage engine btw.

**Comment by dbu** [comment]
on phpcr level, we can't know in advance what type a field is, so that would not be possible
but with the odm query builder we do know the document definition, so we should be able to do the necessary casts. this would need to happen in ConverterPhpcr - but i did not find where exactly, as it is quite generically written. if you want to work on this, ping dantleech, he might have some more specific ideas.

**Comment by steffenbrem** [comment]
@dantleech ping

Thanks for pointing that out, I will try to figure out how the code is flowing in the query builder and try to come up with a solution. If @dantleech could share his thoughts on this, that would be great! :+1:

**Comment by dantleech** [comment]
hi @steffenbrem 

I think a start would be to override the [walkOperandStaticLiteral](https://github.com/doctrine/phpcr-odm/blob/master/lib/Doctrine/ODM/PHPCR/Query/Builder/ConverterBase.php#L525) method from the `ConverterBase` class in the `ConverterPhpcr` class - where you will have access to the document metadata.

Would that make sense @dbu ?

**Comment by dbu** [comment]
sounds reasonable, yes. i was trying to find where i have the metadata,
did not realize the converter base class / phpcr class separation...

**Comment by steffenbrem** [comment]
How would you get the document metadata from that method? I don't see how I could get it.

**Comment by steffenbrem** [comment]
I've fiddled around and came up with the following, is something like this acceptable?

```php
protected function walkOperandStaticLiteral(OperandStaticLiteral $node)
{
    $value = $node->getValue();

    // get the dynamic operand from ConstraintComparison
    $field = $node->getParent()->getChildOfType('operand_dynamic');
    $meta = $this->aliasMetadata[$field->getAlias()];

    // it it's a field/property force cast it to it's type
    if ($field instanceof OperandDynamicField) {
        $field = $field->getField();
        $type = $meta->getFieldMapping($field)['type'];
        settype($value, $type);
    }

    $operand = $this->qomf()->literal($value);
    return $operand;
}
```

**Comment by dantleech** [comment]
Looks generally good to me :) Can you make a PR?

**Comment by steffenbrem** [comment]
@dantleech I've opened a PR for it, I added a type mapping to better isolate this functionality.

**Comment by dbu** [comment]
fixed in #690
